### PR TITLE
`Makefile`, `.ci/write-dockerfile.sh`: Update for `src/pyproject.toml` after #36982

### PR DESCRIPTION
**Pull request by mkoeppe**
Also some unrelated fixes:
- fix for `centos-7-devtoolset` needed to adjust for a small change in how Docker files are parsed (https://github.com/sagemath/sage/actions/runs/8931531493/job/24538795775).
- fix for `ubuntu-xenial-...` and other platforms that need special `configure` args
- fix for current macOS Docker Desktop 

### :memo: Checklist

- [x] The title is concise and informative.
- [ ] The description explains in detail what this PR is about.
- [x] I have linked a relevant issue or discussion.
- [ ] I have created tests covering the changes.
- [ ] I have updated the documentation and checked the documentation preview.

### CODE REVIEW

**Comment by fchapoton** [comment]
hmm, tagging doctests as "known bugs" does not look like a good permanent solution

**Comment by mkoeppe** [comment]
That's right, this is a bad temporary solution.

This is fallout from the big revert in #37796.

**Comment by mkoeppe** [comment]
I've removed the commits that made changes to `.github/workflows/*macos.yml` to reduce the scope of the PR; these commits are now on #37919.

File: .ci/retrofit-worktree.sh [hunk 1]
```diff
@@ -36,6 +38,10 @@ git tag -f new
 # The changed files now show up as uncommitted changes.
 # The final "git add -N" makes sure that files that were added in "new" do not show
 # as untracked files, which would be removed by "git clean -fx".
+if [ -L $WORKTREE_NAME ]; then
+    rm -f $WORKTREE_NAME
+    git worktree prune --verbose
+fi
 git worktree add --detach $WORKTREE_NAME
```
Comment (kwankyu) [inline .ci/retrofit-worktree.sh:45]: Did #36982 make these four lines necessary? I wonder how...

Comment (kwankyu) [reply .ci/retrofit-worktree.sh:45]: Or this "retrofit-workreee" could be applied twice somewhere?

Comment (mkoeppe) [reply .ci/retrofit-worktree.sh:45]: Yes, in Build&Test it's applied three times: 1 time already in the Docker image that it's starting from, and then 2 more times.
This PR is good enough to fix the "CI Linux" workflow that runs on release tags. For "Build&Test", more fixing related to this is necessary.

Comment (kwankyu) [reply .ci/retrofit-worktree.sh:45]: Anyway, it is just cleaning up before actual work.

**Review by kwankyu — APPROVED** [review]

**Pull request merged**